akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -842,61 +842,80 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
-    buffer = BufferImpl(size, ctx.materializer)
-  }
+    var buffer: BufferImpl[T] = _
 
-  override def onPush(elem: T, ctx: DetachedContext[T]): UpstreamDirective =
-    if (ctx.isHoldingDownstream) ctx.pushAndPull(elem)
-    else enqueueAction(ctx, elem)
-
-  override def onPull(ctx: DetachedContext[T]): DownstreamDirective = {
-    if (ctx.isFinishing) {
-      val elem = buffer.dequeue()
-      if (buffer.isEmpty) ctx.pushAndFinish(elem)
-      else ctx.push(elem)
-    } else if (ctx.isHoldingUpstream) ctx.pushAndPull(buffer.dequeue())
-    else if (buffer.isEmpty) ctx.holdDownstream()
-    else ctx.push(buffer.dequeue())
-  }
-
-  override def onUpstreamFinish(ctx: DetachedContext[T]): TerminationDirective =
-    if (buffer.isEmpty) ctx.finish()
-    else ctx.absorbTermination()
-
-  val enqueueAction: (DetachedContext[T], T) ⇒ UpstreamDirective =
-    overflowStrategy match {
-      case DropHead ⇒ (ctx, elem) ⇒
-        if (buffer.isFull) buffer.dropHead()
-        buffer.enqueue(elem)
-        ctx.pull()
-      case DropTail ⇒ (ctx, elem) ⇒
-        if (buffer.isFull) buffer.dropTail()
-        buffer.enqueue(elem)
-        ctx.pull()
-      case DropBuffer ⇒ (ctx, elem) ⇒
-        if (buffer.isFull) buffer.clear()
-        buffer.enqueue(elem)
-        ctx.pull()
-      case DropNew ⇒ (ctx, elem) ⇒
-        if (!buffer.isFull) buffer.enqueue(elem)
-        ctx.pull()
-      case Backpressure ⇒ (ctx, elem) ⇒
-        buffer.enqueue(elem)
-        if (buffer.isFull) ctx.holdUpstream()
-        else ctx.pull()
-      case Fail ⇒ (ctx, elem) ⇒
-        if (buffer.isFull) ctx.fail(new BufferOverflowException(s"Buffer overflow (max capacity was: $size)!"))
-        else {
+    val enqueueAction: T ⇒ Unit =
+      overflowStrategy match {
+        case DropHead ⇒ elem ⇒
+          if (buffer.isFull) buffer.dropHead()
+          buffer.enqueue(elem)
+          pull(in)
+        case DropTail ⇒ elem ⇒
+          if (buffer.isFull) buffer.dropTail()
+          buffer.enqueue(elem)
+          pull(in)
+        case DropBuffer ⇒ elem ⇒
+          if (buffer.isFull) buffer.clear()
+          buffer.enqueue(elem)
+          pull(in)
+        case DropNew ⇒ elem ⇒
+          if (!buffer.isFull) buffer.enqueue(elem)
+          pull(in)
+        case Backpressure ⇒ elem ⇒
           buffer.enqueue(elem)
-          ctx.pull()
+          if (!buffer.isFull) pull(in)
+        case Fail ⇒ elem ⇒
+          if (buffer.isFull) failStage(new BufferOverflowException(s"Buffer overflow (max capacity was: $size)!"))
+          else {
+            buffer.enqueue(elem)
+            pull(in)
+          }
+      }
+
+    override def preStart(): Unit = {
+      buffer = BufferImpl(size, materializer)
+      pull(in)
+    }
+
+    override def onPush(): Unit = {
+      val elem = grab(in)
+      if (isAvailable(out)) {
+        if (buffer.isEmpty) {
+          push(out, elem)
+          pull(in)
+        } else {
+          push(out, buffer.dequeue())
```

Comment:
`isAvailable(out)` can be true here when it has been pulled but there is no element in buffer to delivered. I have written some comments in the source file.